### PR TITLE
Implement custom ETW module support

### PR DESCRIPTION
**Pull request by mjsabby**
Fixes https://github.com/Microsoft/perfview/issues/919.

### CODE REVIEW

**Comment by mjsabby** [comment]
The scenario here is that we have a file format that is a list of EVENT_RECORDs serialized in a special way. So it's very much like ETL, but has a few optimizations. All our tooling is built on TraceEvent, and we'd like to be able to use it to analyze these files.

File: src/TraceEvent/ETWTraceEventSource.cs [hunk 1]
```diff
@@ -383,6 +386,23 @@ public ImageData(int ProcessID, string DllName, Address BaseAddress, int Size)
             public int Size;
         }
 
+        private IntPtr GetFunctionPointerForEtwLibrary(string procName)
+        {
+            IntPtr loadLibraryHandle = TraceEventNativeMethods.LoadLibrary(this.etwModulePath);
+            if (loadLibraryHandle == IntPtr.Zero)
+            {
+                throw new Exception($"LoadLibrary failed. Please check if the file (${this.etwModulePath}) exists, and is of the correct bitness.");
```
Comment (pharring) [inline src/TraceEvent/ETWTraceEventSource.cs]: Use DllNotFoundException.
Also, you have string interpolation inside string interpolation. Remove the inner $

File: src/TraceEvent/ETWTraceEventSource.cs [hunk 2]
```diff
@@ -449,7 +469,17 @@ private void Initialize(string fileOrSessionName, TraceEventSourceType type)
...
             {
-                handles[i] = TraceEventNativeMethods.OpenTrace(ref logFiles[i]);
+                if (string.IsNullOrEmpty(this.etwModulePath))
+                {
+                    handles[i] = TraceEventNativeMethods.OpenTrace(ref logFiles[i]);
+                }
+                else
+                {
+                    var procAddress = this.GetFunctionPointerForEtwLibrary("OpenTraceW");
+                    var openTrace = (TraceEventNativeMethods.OpenTraceDelegate)Marshal.GetDelegateForFunctionPointer(procAddress, typeof(TraceEventNativeMethods.OpenTraceDelegate));
```
Comment (pharring) [inline src/TraceEvent/ETWTraceEventSource.cs]: You should cache this delegate in a field so you don't have to LoadLibrary+GetProcAddress each time around the loop.

**Comment by mjsabby** [comment]
Thanks @pharring. Updated.

**Comment by brianrob** [comment]
@mjsabby thanks for the PR.  I've spent some time thinking about this, and I'd like to propose an alternative design here.  Specifically, rather than building this into TraceEvent, I'd like to propose an interface such as IETWTraceProcessor, an instance of which can be passed to ETWTraceEventSource instead of the 3 new parameters you pass in this PR.  The interface can have methods for opening, closing, and processing the trace just like you do here. - This does a few things:

1. Formalizes the interface in managed code.
2. Allows for more flexibility in the custom handler should you need to do something else (you don't necessarily have to change TraceEvent).
3. You can plumb the start and end time directly through to your custom module instead of having to have TraceEvent know about it (and any future parameters you might like to add).

What do you think?

File: src/TraceEvent/TraceEventNativeMethods.cs [hunk 3]
```diff
@@ -464,6 +468,16 @@ internal unsafe struct EVENT_FILTER_EVENT_ID
 
         #endregion
 
+        #region Library functions
+
+        [DllImport("kernel32.dll", CharSet = CharSet.Unicode)]
+        internal static extern IntPtr LoadLibrary([MarshalAs(UnmanagedType.LPWStr)]string lpFileName);
+
+        [DllImport("kernel32.dll", CharSet = CharSet.Unicode)]
```
Comment (pharring) [inline src/TraceEvent/TraceEventNativeMethods.cs]: There is no GetProcAddressW. It's ANSI only. So it should be `Charset.ANSI`, `UnmanagedType.LPStr` and you probably want to add `ExactSpelling = true` for good measure (so the CLR doesn't try to find the A/W version)

**Comment by mjsabby** [comment]
@brianrob Sounds good to me.

**Pull request closed without merging**